File: crates/tsgo_rs/examples/mock_client.rs
mod support;

use serde_json::json;
use tsgo_rs::{
    TsgoError,
    api::{ApiClient, ApiMode, UpdateSnapshotParams},
    runtime::block_on,
};

fn main() -> Result<(), tsgo_rs::TsgoError> {
    let result = block_on(async {
        let client = ApiClient::spawn(support::mock_api_config(
            "mock_client",
            ApiMode::AsyncJsonRpcStdio,
        )?)
        .await?;
        let init = client.initialize().await?;
        let config = client.parse_config_file("/workspace/tsconfig.json").await?;
        let snapshot = client
            .update_snapshot(UpdateSnapshotParams {
                open_project: Some("/workspace/tsconfig.json".into()),
                file_changes: None,
            })
            .await?;
        let project = snapshot.projects.first().ok_or_else(|| {
            TsgoError::Protocol("mock client example did not return a project".into())
        })?;
        let source = client
            .get_source_file(
                snapshot.handle.clone(),
                project.id.clone(),
                "/workspace/src/index.ts",
            )
            .await?
            .ok_or_else(|| {
                TsgoError::Protocol("mock client example did not return source".into())
            })?;
        let source_text = String::from_utf8(source.as_bytes().to_vec()).map_err(|err| {
            TsgoError::Protocol(format!("source file was not utf8: {err}").into())
        })?;
        let string_type = client
            .get_string_type(snapshot.handle.clone(), project.id.clone())
            .await?;
        let rendered = client
            .type_to_string(
                snapshot.handle.clone(),
                project.id.clone(),
                string_type.id.clone(),
                None,
                None,
            )
            .await?;

        let result = json!({
            "currentDirectory": init.current_directory,
            "projectId": project.id,
            "configFileName": project.config_file_name,
            "fileCount": config.file_names.len(),
            "sourceFileText": source_text,
            "stringTypeText": rendered,
        });
        snapshot.release().await?;
        client.close().await?;
        Ok::<_, tsgo_rs::TsgoError>(result)
    })?;

    support::print_json(result);
    Ok(())
}

File: examples/nodejs/mock_client.ts
import { TsgoApiClient } from "@corsa-bind/node";

import { assertExists, isMain, mockBinary, workspaceRoot } from "../shared.ts";

export interface MockClientExampleResult {
  currentDirectory: string;
  projectId: string;
  sourceFileText: string;
  stringTypeText: string;
}

export function runMockClientExample(): MockClientExampleResult {
  assertExists(
    mockBinary,
    "mock tsgo binary",
    "run `vp run -w build_mock` or `vp run -w build` first",
  );

  const client = TsgoApiClient.spawn({
    executable: mockBinary,
    cwd: workspaceRoot,
    mode: "jsonrpc",
  });

  let snapshotHandle: string | undefined;

  try {
    const init = client.initialize();
    const snapshot = client.updateSnapshot({
      openProject: "/workspace/tsconfig.json",
    });
    snapshotHandle = snapshot.snapshot;
    const project = snapshot.projects[0];
    if (!project) {
      throw new Error("mock client did not return a project");
    }

    const sourceFile = client.getSourceFile(
      snapshot.snapshot,
      project.id,
      "/workspace/src/index.ts",
    );
    const stringType = client.getStringType(snapshot.snapshot, project.id);

    return {
      currentDirectory: init.currentDirectory,
      projectId: project.id,
      sourceFileText: Buffer.from(sourceFile ?? []).toString("utf8"),
      stringTypeText: client.typeToString(snapshot.snapshot, project.id, stringType.id),
    };
  } finally {
    if (snapshotHandle) {
      client.releaseHandle(snapshotHandle);
    }
    client.close();
  }
}

if (isMain(import.meta.url)) {
  console.log(JSON.stringify(runMockClientExample(), null, 2));
}
